src/ray/raylet/scheduling/cluster_task_manager_test.cc

```diff
@@ -0,0 +1,343 @@
+  JobID job_id = RandomJobId();
+  rpc::Address address;
+  spec_builder.SetCommonTaskSpec(
+      id, Language::PYTHON, FunctionDescriptorBuilder::BuildPython("", "", "", ""),
+      job_id, TaskID::Nil(), 0, TaskID::Nil(), address, 0, required_resources, {});
+  rpc::TaskExecutionSpec execution_spec_message;
+  execution_spec_message.set_num_forwards(1);
+  return Task(spec_builder.Build(), TaskExecutionSpecification(execution_spec_message));
+}
+
+class ClusterTaskManagerTest : public ::testing::Test {
+ public:
+  ClusterTaskManagerTest()
+      : id_(ClientID::FromRandom()),
+        single_node_resource_scheduler_(CreateSingleNodeScheduler(id_.Binary())),
+        fulfills_dependencies_calls_(0),
+        dependencies_fulfilled_(true),
+        node_info_calls_(0),
+        node_info_(boost::optional<rpc::GcsNodeInfo>{}),
+        task_manager_(id_, single_node_resource_scheduler_,
+                      [this](const Task &_task) {
+                        fulfills_dependencies_calls_++;
+                        return dependencies_fulfilled_;
+                      },
+                      [this](const ClientID &node_id) {
+                        node_info_calls_++;
+                        return node_info_;
+                      }) {}
+
+  void SetUp() {}
+
+  void Shutdown() {}
+
+  ClientID id_;
+  std::shared_ptr<ClusterResourceScheduler> single_node_resource_scheduler_;
+  MockWorkerPool pool_;
+  std::unordered_map<WorkerID, std::shared_ptr<WorkerInterface>> leased_workers_;
+
+  int fulfills_dependencies_calls_;
+  bool dependencies_fulfilled_;
+
+  int node_info_calls_;
+  boost::optional<rpc::GcsNodeInfo> node_info_;
+
+  ClusterTaskManager task_manager_;
+};
+
+TEST_F(ClusterTaskManagerTest, BasicTest) {
+  /*
+    Test basic scheduler functionality:
+    1. Queue and attempt to schedule/dispatch atest with no workers available
+    2. A worker becomes available, dispatch again.
+   */
+  Task task = CreateTask({{ray::kCPU_ResourceLabel, 4}});
+  rpc::RequestWorkerLeaseReply reply;
+  bool callback_occurred = false;
+  bool *callback_occurred_ptr = &callback_occurred;
+  auto callback = [callback_occurred_ptr]() { *callback_occurred_ptr = true; };
+
+  task_manager_.QueueTask(task, &reply, callback);
+  task_manager_.SchedulePendingTasks();
+  task_manager_.DispatchScheduledTasksToWorkers(pool_, leased_workers_);
+
+  ASSERT_FALSE(callback_occurred);
+  ASSERT_TRUE(leased_workers_.size() == 0);
+  ASSERT_TRUE(pool_.workers.size() == 0);
+
+  std::shared_ptr<MockWorker> worker =
+      std::make_shared<MockWorker>(WorkerID::FromRandom(), 1234);
+  pool_.PushWorker(std::dynamic_pointer_cast<WorkerInterface>(worker));
+
+  task_manager_.DispatchScheduledTasksToWorkers(pool_, leased_workers_);
+
+  ASSERT_TRUE(callback_occurred == true);
+  ASSERT_TRUE(leased_workers_.size() == 1);
+  ASSERT_TRUE(pool_.workers.size() == 0);
+  ASSERT_TRUE(fulfills_dependencies_calls_ == 0);
+  ASSERT_TRUE(node_info_calls_ == 0);
```

Comment:
You can also use `ASSERT_EQ`

Reply:
ASSERT_EQ behaves unintuitively (really arguable incorrectly) on some objects, so I think it's easier to just always use ASSER_TRUE (see the cluster_resource_scheduler_tests)

Reply:
Hmm what sort of objects have you seen that for? I actually think it's a bit nicer since it prints the actual value on assertion failure, but it was just a suggestion.

Reply:
i swear i have seen this happen before but figure out where right now. changing this to EQ seems totally reasonable right now, and i'll let you know when it becomes a problem again